numba/tests/test_array_exprs.py

```diff
@@ -0,0 +1,252 @@
+from __future__ import print_function, division, absolute_import
+from timeit import Timer
+
+import numpy as np
+
+from numba import njit, vectorize
+from numba import unittest_support as unittest
+from numba import compiler, typing, typeof, ir
+from numba.compiler import Pipeline, _PipelineManager, Flags
+from numba.targets import cpu
+
+
+class Namespace(dict):
+    def __getattr__(s, k):
+        return s[k] if k in s else super(Namespace, s).__getattr__(k)
+
+def axy(a, x, y):
+    return a * x + y
+
+def ax2(a, x, y):
+    return a * x + y
+
+def pos_root(As, Bs, Cs):
+    return (-Bs + (((Bs ** 2.) - (4. * As * Cs)) ** 0.5)) / (2. * As)
+
+def neg_root_common_subexpr(As, Bs, Cs):
+    _2As = 2. * As
+    _4AsCs = 2. * _2As * Cs
+    _Bs2_4AsCs = (Bs ** 2. - _4AsCs)
+    return (-Bs - (_Bs2_4AsCs ** 0.5)) / _2As
+
+def neg_root_complex_subexpr(As, Bs, Cs):
+    _2As = 2. * As
+    _4AsCs = 2. * _2As * Cs
+    _Bs2_4AsCs = (Bs ** 2. - _4AsCs) + 0j # Force into the complex domain.
+    return (-Bs - (_Bs2_4AsCs ** 0.5)) / _2As
+
+vaxy = vectorize(axy)
+
+def call_stuff(a0, a1):
+    return np.cos(vaxy(a0, np.sin(a1) - 1., 1.))
+
+def are_roots_imaginary(As, Bs, Cs):
+    return (Bs ** 2 - 4 * As * Cs) < 0
+
+
+class RewritesTester(Pipeline):
+    @classmethod
+    def mk_pipeline(cls, args, return_type=None, flags=None, locals={},
+                    library=None, typing_context=None, target_context=None):
+        if not flags:
+            flags = Flags()
+        flags.nrt = True
+        if typing_context is None:
+            typing_context = typing.Context()
+        if target_context is None:
+            target_context = cpu.CPUContext(typing_context)
+        return cls(typing_context, target_context, library, args, return_type,
+                   flags, locals)
+
+    @classmethod
+    def mk_no_rw_pipeline(cls, args, return_type=None, flags=None, locals={},
+                          library=None, **kws):
+        if not flags:
+            flags = Flags()
+        flags.no_rewrites = True
+        return cls.mk_pipeline(args, return_type, flags, locals, library, **kws)
+
+
+class TestArrayExpressions(unittest.TestCase):
+
+    def test_simple_expr(self):
+        '''
+        Using a simple array expression, verify that rewriting is taking
+        place, and is fusing loops.
+        '''
+        A = np.linspace(0,1,10)
+        X = np.linspace(2,1,10)
+        Y = np.linspace(1,2,10)
+        arg_tys = [typeof(arg) for arg in (A, X, Y)]
+
+        control_pipeline = RewritesTester.mk_no_rw_pipeline(arg_tys)
+        cres_0 = control_pipeline.compile_extra(axy)
+        nb_axy_0 = cres_0.entry_point
+
+        test_pipeline = RewritesTester.mk_pipeline(arg_tys)
+        cres_1 = test_pipeline.compile_extra(axy)
+        nb_axy_1 = cres_1.entry_point
+
+        control_pipeline2 = RewritesTester.mk_no_rw_pipeline(arg_tys)
+        cres_2 = control_pipeline2.compile_extra(ax2)
```

Comment:
Just being super paranoid; one of the pipelines might have altered axy, and I wanted a "second opinion" at hand.